Cf 1805 f1 (#6360)
cpinitiative/usaco-guide: solutions/silver/cf-1805F1.mdx

Sadly, we can't repeatedly apply $F$ on the raw values: we need the true relative order of elements to run the k-way merge correctly, so we can't reduce anything mod $p$ until the very end, but the values themselves grow enormously over $n - 1$ rounds. This means that we must normalise the array in some way.

We notice that $F$ works nicely with constant shifts: if each element of an array is shifted by $c$, every pairwise sum shifts by $2c$. So instead of storing the true array directly, maintain it as $\text{cur}[i] + \text{offset}$ for a shared scalar $\text{offset}$. Applying $F$ to the true array is the same as applying it to $\text{cur}$ alone and adding $2 \cdot \text{offset}$ to the result. If we stored this new array directly, its values would still continue to grow after every application of $F$. Thus, we normalize it by subtracting its minimum value. After computing the merged array $\text{nxt}$, let $m = \text{nxt}[0]$ be its minimum, we update

> $$\text{offset} \leftarrow 2 \cdot \text{offset} + m, \qquad \text{cur}[k] \leftarrow \text{nxt}[k] - m.$$

Reviewer: ```suggestion
$$
\text{offset} \leftarrow 2 \cdot \text{offset} + m, \qquad \text{cur}[k] \leftarrow \text{nxt}[k] - m.
$$
```

Since only $\text{cur}$ is used for comparisons, $\text{offset}$ can be reduced mod $p$ immediately. Meanwhile $\text{cur}$ always has minimum $0$ and its entries stay bounded by the sum of two elements, so it never overflows. The true final answer is $(\text{offset} + \text{cur}[0]) \bmod p$.

## Implementation

One application of $F$ costs $\mathcal{O}(n \log n)$, and array size shrinks by $1$ each round for $n - 1$ rounds, giving an overall complexity of $\mathcal{O}(n^2 \log n)$.

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

> const long long MOD = 1e9 + 7;

Reviewer: ```suggestion
const int MOD = 1e9 + 7;
```

</Spoiler>

<Spoiler title="Hint 2">

By fixing the left endpoint $i$ of a pair - the sums $a_i + a_{i+1}, a_i + a_{i+2}, a_i + a_{i+3}, \ldots$ come out sorted too. So you now have $n - 1$ separately sorted lists (one per left endpoint). How can we find the smallest $n - 1$ elements across several sorted lists?

</Spoiler>

<Spoiler title="Hint 3">

Applying $F$ many times on the raw values will overflow, but you can't reduce mod $p$ mid-way either, since comparisons need the true order. Can you track the array as a small, bounded part plus a separate scalar that's allowed to be reduced mod $p$?

</Spoiler>

## Explanation

Sort the array, then fix a left endpoint $i$ of a pair. The sums $a_i + a_{i+1}, a_i + a_{i+2}, a_i + a_{i+3}, \ldots$ are in increasing order, since the array is sorted. So now instead of one big set of $\mathcal{O}(n^2)$ sums, we have $n - 1$ individually sorted lists, one per left endpoint, and finding the smallest $n - 1$ elements across all of them is just a **k-way merge**.

> We can maintain a priority queue for holding the smallest unused element of every list, i.e. for every $i$ we initially push $(a_i + a_{i+1}, i, i+1)$. We repeatedly pop the smallest sum, append it to the output, and push the next element of that same list, $(a_i + a_{j+1}, i, j+1)$, if it exists. Since each list is individually sorted, the priority queue will always hold the smallest unused sum from each, and the merged output will be globally sorted. Each push/pop is $\mathcal{O}(\log n)$ and we do $n - 1$ of them, so one application of $F$ costs $\mathcal{O}(n \log n)$.

Reviewer: You can modify these paragraph. like firstly we do this, secondly this and ......

Sadly, we can't repeatedly apply $F$ on the raw values: we need the true relative order of elements to run the k-way merge correctly, so we can't reduce anything mod $p$ until the very end, but the values themselves grow enormously over $n - 1$ rounds. This means that we must normalise the array in some way.

We notice that $F$ works nicely with constant shifts: if each element of an array is shifted by $c$, every pairwise sum shifts by $2c$. So instead of storing the true array directly, maintain it as $\text{cur}[i] + \text{offset}$ for a shared scalar $\text{offset}$. Applying $F$ to the true array is the same as applying it to $\text{cur}$ alone and adding $2 \cdot \text{offset}$ to the result. If we stored this new array directly, its values would still continue to grow after every application of $F$. Thus, we normalize it by subtracting its minimum value. After computing the merged array $\text{nxt}$, let $m = \text{nxt}[0]$ be its minimum, we update

$$
\text{offset} \leftarrow 2 \cdot \text{offset} + m, \qquad \text{cur}[k] \leftarrow \text{nxt}[k] - m.
$$

> Since only $\text{cur}$ is used for comparisons, $\text{offset}$ can be reduced mod $p$ immediately. Meanwhile $\text{cur}$ always has minimum $0$ and its entries stay bounded by the sum of two elements, so it never overflows. The true final answer is $(\text{offset} + \text{cur}[0]) \bmod p$.

Reviewer: Which two elements?

## Implementation

> Each application of $F$ costs $\mathcal{O}(n \log n)$, giving an overall complexity of $\mathcal{O}(n^2 \log n)$.

Reviewer: just mention total tc, you have already explained about application cost in edi

$$
\text{offset} \leftarrow 2 \cdot \text{offset} + m, \qquad \text{cur}[k] \leftarrow \text{nxt}[k] - m.
$$

Since only $\text{cur}$ is used for comparisons, $\text{offset}$ can be reduced mod $p$ immediately. Meanwhile, after subtracting the minimum value in each round, $\text{cur}$ always has minimum $0$, so its entries never overflow.The true final answer is $(\text{offset} + \text{cur}[0]) \bmod p$.

## Implementation

**Time Complexity:** $\mathcal{O}(n^2 \log n)$.

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

const int MOD = 1e9 + 7;

int main() {
	ios::sync_with_stdio(false);
	cin.tie(nullptr);

	int n;
	cin >> n;

	vector<long long> cur(n);
	for (auto &x : cur) cin >> x;
	sort(cur.begin(), cur.end());

	long long offset = 0;

	while (cur.size() > 1) {
		int m = cur.size();

		// Min-heap of (pair sum, left endpoint, right endpoint)
		priority_queue<tuple<long long, int, int>, vector<tuple<long long, int, int>>,
		               greater<>>

> 		    pq;

Reviewer: formatting is a bit off, use #define something for tuple<long long, int, int>

Reviewer 2: i would either do smth like

```cpp
using state = std::tuple<long long, int, int>;
```

or make a struct

(also ngl i would always personally just use `std::array<long long, 3>` because i'm lazy lmao, but tuple is fine ig)

Reviewer 3: Yeah I think we should definitely switch to structs here all the way through, a struct for a heap item, left endpoint, and right endpoint imho would be the best as you can access things in a semantic manner but keeping tuple is fine.  @4nav this is down to what you prefer but at least do a typedef but consider a struct

Author: gotcha

We notice that $F$ works nicely with constant shifts: if each element of an array is shifted by $c$, every pairwise sum shifts by $2c$. So instead of storing the true array directly, maintain it as $\text{cur}[i] + \text{offset}$ for a shared scalar $\text{offset}$. Applying $F$ to the true array is the same as applying it to $\text{cur}$ alone and adding $2 \cdot \text{offset}$ to the result. If we stored this new array directly, its values would still continue to grow after every application of $F$. Thus, we normalize it by subtracting its minimum value. After computing the merged array $\text{nxt}$, let $m = \text{nxt}[0]$ be its minimum, we update 

$$\text{offset} \leftarrow 2 \cdot \text{offset} + m, \qquad \text{cur}[k] \leftarrow \text{nxt}[k] - m.$$
 
Since only $\text{cur}$ is used for comparisons, $\text{offset}$ can be reduced mod $p$ immediately. Meanwhile $\text{cur}$ always has minimum $0$ and its entries stay bounded by the sum of two elements, so it never overflows. The true final answer is $(\text{offset} + \text{cur}[0]) \bmod p$.

## Implementation
 

> One application of $F$ costs $\mathcal{O}(n \log n)$, and array size shrinks by $1$ each round for $n - 1$ rounds, giving an overall complexity of $\mathcal{O}(n^2 \log n)$.

Reviewer: just mention the total tc

 
<LanguageSection>
<CPPSection>
  
```cpp
#include <bits/stdc++.h>
using namespace std;
 
const long long MOD = 1e9 + 7;
 
int main() {
	ios::sync_with_stdio(false);
	cin.tie(nullptr);
 
	int n;
	cin >> n;
 
	vector<long long> cur(n);
	for (auto &x : cur) cin >> x;
	sort(cur.begin(), cur.end());
 
	long long offset = 0;

>  
> 	while (cur.size() > 1) {
> 		int m = cur.size();
>  
> 		// Min-heap of (pair sum, left endpoint, right endpoint)
> 		priority_queue<tuple<long long, int, int>, vector<tuple<long long, int, int>>,
> 		    greater<>> pq;
>  
> 		for (int i = 0; i < m - 1; i++) {
> 			pq.push({cur[i] + cur[i + 1], i, i + 1});
> 		}
>  
> 		// k-way merge: pull the smallest n-1 sums across all m-1 sorted lists
> 		vector<long long> nxt;
> 		while ((int)nxt.size() < m - 1) {
> 			auto [sum, i, j] = pq.top();
> 			pq.pop();
> 			nxt.push_back(sum);
> 			if (j + 1 < m) pq.push({cur[i] + cur[j + 1], i, j + 1});
> 		}
>  
> 		// Fold the common minimum into offset (mod p) to keep cur small
> 		long long min_sum = nxt[0];
> 		offset = (2 * offset + min_sum) % MOD;
> 		for (auto &x : nxt) x -= min_sum;
>  
> 		cur = move(nxt);
> 	}
>  
> 	cout << (offset + cur[0]) % MOD << "\n";
> }
> ```
> 
> </CPPSection>
> </LanguageSection>
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 

Reviewer: ```suggestion
```

---
id: cf-1805F1
source: CF
title: Survival of the Weakest (easy version)
author: Abhinav Singh
---

## Hints

<Spoiler title="Hint 1">

Computing all $\mathcal{O}(n^2)$ pair sums directly is too slow to repeat $n - 1$ times.
Try sorting the array: does that reveal any structure in how the pair sums compare to each other?

</Spoiler>

<Spoiler title="Hint 2">

By fixing the left endpoint $i$ of a pair - the sums $a_i + a_{i+1}, a_i + a_{i+2}, a_i + a_{i+3}, \ldots$
come out sorted too. So you now have $n - 1$ separately sorted lists (one per left endpoint).

> How can we find the smallest $n - 1$ elements across several sorted lists?

Reviewer 2: idk if i would think of it as separate ordered lists

for me, my thought process was like, if we fix left endpoint $i$ and have used up all right endpoints up to some $j$, then we obviously don't need to immediately consider $(i, j + 1)$

so that means for each left endpoint we consider exactly 1 possible pair in a stage of our algo until we reach n - 1 elements

Author: thats like saying the same thing almost tho right

Reviewer 2: sorta, it's just how u think abt it

this is more subjective, if any other reviewer wants to weigh in feel free

Author: @johnzhou721 @Sosuke23

Reviewer 3: I think the way eys thought about it is more intuitive, we should be explicit instead of using some metaphor about lists

Author: hmmm  sure

are in increasing order, since the array is sorted. So now instead of one big set of $\mathcal{O}(n^2)$ sums,
we have $n - 1$ individually sorted lists, one per left endpoint, and our problem is reduced to finding
the smallest $n - 1$ elements across all of them.

Now, we can maintain a priority queue containing the smallest unused element from each list. Initially,
for every $i$, push $(a_i + a_{i+1}, i, i+1)$ into the queue. We repeatedly pop the smallest sum,
append it to the merged array, and push the next element of that same list, $(a_i + a_{j+1}, i, j+1)$, if it exists.

Since each list is individually sorted, the priority queue will always hold the smallest unused sum
from each list. Thereforce, each extraction is the smallest remaining pair **overall**, so the merged
output will be globally sorted. Each push/pop is $\mathcal{O}(\log n)$ and we do $n - 1$ of them,
so one application of $F$ costs $\mathcal{O}(n \log n)$.

Sadly, we can't repeatedly apply $F$ on the raw values: we need the true relative order of elements to
run the k-way merge correctly, so we can't reduce anything mod $p$ until the very end, but the values

> themselves grow enormously over $n - 1$ rounds. This means that we must normalise the array in some way.

Reviewer 2: should explain why it's necessary to reduce the numbers somehow (ie since it's possible values will keep doubling)

We notice that $F$ works nicely with constant shifts: if each element of an array is shifted by $c$,
every pairwise sum shifts by $2c$. So instead of storing the true array directly, maintain it as

> $(\text{cur}[i] + \text{offset})$ for a shared scalar $\text{offset}$. Applying $F$ to the true array

Reviewer 2: what is $cur$? the array with the offset applied?

Author: yeah

Author: will fix

is the same as applying it to $\text{cur}$ alone and adding $(2 \cdot \text{offset})$ to the result.
If we stored this new array directly, its values would still continue to grow after every application
of $F$. Thus, we normalize it by subtracting its minimum value. After computing the merged
array $\text{nxt}$, let $m = \text{nxt}[0]$ be its minimum, we update

$$
\text{offset} \leftarrow 2 \cdot \text{offset} + m, \qquad \text{cur}[k] \leftarrow \text{nxt}[k] - m.
$$

> Since only $\text{cur}$ is used for comparisons, $\text{offset}$ can be reduced mod $p$ immediately.

Reviewer 2: imo, this doesn't really explain why it can't overflow?

if we let the max value in an array $a$ be $r$, then the min value must be $\geq r$. the max pairwise value is $2r$, so max - min is $\leq r$. since $a_i < 10^9$, values will always be less than $10^9$

Author: by min i believe you mean minimum pairwise sum in the pq but why is it necessarily >=r? its gonna be >= the minimum element in the array, not the maximum?

Reviewer 2: oh shoot i was being dumb 💀 i forgot the problem was for minimums

Meanwhile, after subtracting the minimum value in each round, $\text{cur}$ always has minimum $0$,
so its entries never overflow.The true final answer is $(\text{offset} + \text{cur}[0]) \bmod p$.

## Implementation

> **Time Complexity:** $\mathcal{O}(n^2 \log n)$.

Reviewer 2: ```suggestion
**Time Complexity:** $\mathcal{O}(N^2 \log N)$.
```

keep time complexity uppercase

---
id: cf-1805F1
source: CF
title: Survival of the Weakest (easy version)
author: Abhinav Singh
---

## Hints

> <Spoiler title="Hint 1">
> 
> Computing all $\mathcal{O}(n^2)$ pair sums directly is too slow to repeat $n - 1$ times.
> Try sorting the array: does that reveal any structure in how the pair sums compare to each other?
> 
> </Spoiler>
> 
> <Spoiler title="Hint 2">
> 
> For each left endpoint $i$, keep track of the smallest unused pair involving $i$. How can we efficiently find the smallest such pair overall and update its right endpoint?
> 
> </Spoiler>
> 
> <Spoiler title="Hint 3">
> 
> Applying $F$ many times on the raw values will overflow, but you can't reduce mod $p$ mid-way
> either, since comparisons need the true order. Can you track the array as a small,
> bounded part plus a separate scalar that's allowed to be reduced mod $p$?
> 
> </Spoiler>

Reviewer 2: ```suggestion
<Spoiler title="Hint 1">

Computing all $\mathcal{O}(N^2)$ pair sums directly is too slow. Try sorting the array and working out a 
few sample cases. Do you notice any structure in the indices used in the smallest pairwise sums?

</Spoiler>

<Spoiler title="Hint 2">

For each left endpoint $i$, keep track of the smallest unused pair involving $i$.  Across all $i$, how can we efficiently find the smallest such pair, and update it after using a pair?

</Spoiler>

<Spoiler title="Hint 3">

Applying $F$ directly will eventually overflow, since the smallest value in the array will always double. Additionally, we cannot take the remainder after each application of $F$, since comparisons are necessary.

Try to bound the values in our array after each application of $F$, with a separate scalar to recover the true answer at the end.

</Spoiler>
```

not super necessary but imo rewording the hints may be helpful

are in increasing order, since the array is sorted. So now instead of one big set of $\mathcal{O}(n^2)$ sums,
we have $n - 1$ individually sorted lists, one per left endpoint, and our problem is reduced to finding
the smallest $n - 1$ elements across all of them.

Now, we can maintain a priority queue containing the smallest unused element from each list. Initially,
for every $i$, push $(a_i + a_{i+1}, i, i+1)$ into the queue. We repeatedly pop the smallest sum,
append it to the merged array, and push the next element of that same list, $(a_i + a_{j+1}, i, j+1)$, if it exists.

Since each list is individually sorted, the priority queue will always hold the smallest unused sum
from each list. Thereforce, each extraction is the smallest remaining pair **overall**, so the merged
output will be globally sorted. Each push/pop is $\mathcal{O}(\log n)$ and we do $n - 1$ of them,
so one application of $F$ costs $\mathcal{O}(n \log n)$.

Sadly, we can't repeatedly apply $F$ on the raw values: we need the true relative order of elements to
run the k-way merge correctly, so we can't reduce anything mod $p$ until the very end, but the values

> themselves grow enormously over $n - 1$ rounds. In fact, every application of $F$ can nearly double the largest value, so after many rounds

Reviewer 2: > every application of $F$ can nearly double the largest value

i think it's more precise to state that the minimum element will always at least double, which can overflow

the numbers quickly exceed the range of any integer. This means that we must normalise the array in some way.

We notice that $F$ works nicely with constant shifts: if each element of an array is shifted by $c$,
every pairwise sum shifts by $2c$. So instead of storing the true array directly, maintain it as
$(\text{cur}[i] + \text{offset})$, where $\text{cur}$ is a normalized array used for comparisons and
$\text{offset}$ is a shared scalar. Applying $F$ to the true array is the same as applying it to $\text{cur}$
alone and adding $(2 \cdot \text{offset})$ to the result.
If we stored this new array directly, its values would still continue to grow after every application
of $F$. Thus, we normalize it by subtracting its minimum value. After computing the merged
array $\text{nxt}$, let $m = \text{nxt}[0]$ be its minimum, we update

$$
\text{offset} \leftarrow 2 \cdot \text{offset} + m, \qquad \text{cur}[k] \leftarrow \text{nxt}[k] - m.
$$

Since only $\text{cur}$ is used for comparisons, $\text{offset}$ can be reduced mod $p$ immediately.
Meanwhile, after subtracting the minimum value in each round, $\text{cur}$ always has minimum $0$,

> so its entries never overflow.The true final answer is $(\text{offset} + \text{cur}[0]) \bmod p$.

Reviewer 2: 1. IMO the tone should match typical math/cs writing a bit more (mainly, avoid ambiguous phrases like 'anything', more 'emotional' words like 'sadly', and try to keep things concise). 
2. as mentioned earlier, i think explaining it as separate lists is kinda roundabout, since the idea boils down to "hey if it's sorted and $(i, j)$ is smallest unused for left endpoint $i$, we don't need to consider $(i, j+1)$ at this moment unless $(i, j)$ gets used"

for the tone part, it's generally fine if it leans *slightly* more conversational (see [this internal sol](https://usaco.guide/problems/usaco-1323-feb/solution), which is excellent), but it should still be precise and concise

Author: oopsies, my bad i forgot to change the lists idea here i fixed it above but forgot here, will do. Also regarding writing more concisely, what parts should i generally skim over/not overexplain and what parts do u recommend i do? i tend to write an editorial like how im gonna explain the solution while conversing, which tends to go way more in depth than necessary as I've noticed.

Reviewer 2: for this editorial, i think the actual amount of content ur explaining is fine (e.g. i dont think there's anything too overexplained)

it's mainly just cutting down on some 'fluff' words and using more precise language in certain areas (lmk if you want examples)

## Implementation

> **Time Complexity:** $\mathcal{O}(N^2 \log N)$.

Reviewer 2: ```suggestion
**Time Complexity:** $\mathcal{O}(N^2 \log N)$
```
